cmd/gapit/benchmark.go

```diff
@@ -0,0 +1,848 @@
+// Copyright (C) 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"flag"
+	"fmt"
+	"io"
+	"os"
+	"strconv"
+	"strings"
+	"sync"
+	"text/tabwriter"
+	"time"
+
+	"github.com/google/gapid/core/app"
+	"github.com/google/gapid/core/app/status"
+	"github.com/google/gapid/core/event/task"
+	img "github.com/google/gapid/core/image"
+	"github.com/google/gapid/core/log"
+	"github.com/google/gapid/core/os/device"
+	"github.com/google/gapid/gapis/api"
+	"github.com/google/gapid/gapis/client"
+	"github.com/google/gapid/gapis/service"
+	"github.com/google/gapid/gapis/service/path"
+	"github.com/google/gapid/gapis/stringtable"
+)
+
+type benchmarkVerb struct {
+	BenchmarkFlags
+	startTime            time.Time
+	beforeStartTraceTime time.Time
+	traceInitializedTime time.Time
+	traceDoneTime        time.Time
+	traceSizeInBytes     int64
+	traceFrames          int
+	gapisInteractiveTime time.Time
+	gapisCachingDoneTime time.Time
+	interactionStartTime time.Time
+	interactionDoneTime  time.Time
+}
+
+var BenchmarkName = "benchmark.gfxtrace"
+
+func init() {
+	verb := &benchmarkVerb{}
+
+	app.AddVerb(&app.Verb{
+		Name:      "benchmark",
+		ShortHelp: "Runs a set of benchmarking tests on an application",
+		Action:    verb,
+	})
+}
+
+// We wnat to write our some of our own tracing data
+type profileTask struct {
+	Name      string `json:"name,omitempty"`
+	Pid       uint64 `json:"pid"`
+	Tid       uint64 `json:"tid"`
+	EventType string `json:"ph"`
+	Ts        int64  `json:"ts"`
+	S         string `json:"s,omitempty"`
+}
+
+func (verb *benchmarkVerb) Run(ctx context.Context, flags flag.FlagSet) error {
```

Comment:
I know this is effectively 'only' a test, but I find reading functions that are 600 lines+ quite difficult :smile: . It would be nice to have this split up, and sharing code with the rest of gapit, so that future changes are simple and non-duplicated.